### ISSUE TITLE
\i raises a TypeError

### ISSUE THREAD

**Issue by lelit** [dbcli/pgcli#603]
With Python 3.6b2, I got the following immediately after entering `\i` followed by a space:

```
postgres@mydb> \i Exception in thread Thread-4:                            
Traceback (most recent call last):                                              
  File "/usr/local/lib/python3.6/threading.py", line 916, in _bootstrap_inner   
    self.run()                                                                  
  File "/usr/local/lib/python3.6/threading.py", line 864, in run                
    self._target(*self._args, **self._kwargs)                                   
  File "/usr/local/lib/python3.6/site-packages/prompt_toolkit/interface.py", line 841, in run
    completions = list(buffer.completer.get_completions(document, complete_event))                                                                                                                                 
  File "/usr/local/lib/python3.6/site-packages/pgcli/pgcompleter.py", line 371, in get_completions
    reverse=True)                                                               
TypeError: '<' not supported between instances of 'NoneType' and 'NoneType'     
```

I cannot investigate further right now, but I will try to narrow the case.

**Comment by j-bennet** [comment]
Hmm, I can't even get this far. For me, running in 3.6 breaks when importing sqlparse.

```
Traceback (most recent call last):
  File "/Users/irina/.venvs/pgcli36/bin/pgcli", line 11, in <module>
    load_entry_point('pgcli', 'console_scripts', 'pgcli')()
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/pkg_resources/__init__.py", line 564, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2608, in load_entry_point
    return ep.load()
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2268, in load
    return self.resolve()
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2274, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/Users/irina/src/pgcli/pgcli/main.py", line 38, in <module>
    from pgspecial.main import (PGSpecial, NO_QUERY, content_exceeds_width)
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/pgspecial/__init__.py", line 12, in <module>
    from . import iocommands
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/pgspecial/iocommands.py", line 7, in <module>
    import sqlparse
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/sqlparse/__init__.py", line 15, in <module>
    from sqlparse import filters
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/sqlparse/filters/__init__.py", line 16, in <module>
    from sqlparse.filters.tokens import KeywordCaseFilter
  File "/Users/irina/.venvs/pgcli36/lib/python3.6/site-packages/sqlparse/filters/tokens.py", line 26, in <module>
    class KeywordCaseFilter(_CaseFilter):
TypeError: '_TokenType' object is not callable
```

What version of sqlparse do you have installed @lelit ? (so I can maybe move past sqlparse and look at your issue).

**Comment by lelit** [comment]
I'm using Python 3.6b2 and sqlparse 0.2.1.

sqlparse 0.2.2, just released, requires current tip version of Python, due to https://github.com/andialbrecht/sqlparse/issues/286, the fix of which has been reverted in latest sqlparse.

Thank you for looking!

**Comment by j-bennet** [comment]
Finally got through that hurdle. Surprisingly, `\i` works fine:

```
(pgcli36b2) --- src/pgcli ‹master*?› » pgcli boo
Version: 1.1.0
Chat: https://gitter.im/dbcli/pgcli
Mail: https://groups.google.com/forum/#!forum/pgcli
Home: http://pgcli.com
boo> \i ~/data/basic.sql
+------+--------+
|   id | name   |
|------+--------|
|    1 | John   |
|    2 | Mary   |
|    3 | Suzie  |
|    4 | Mike   |
+------+--------+
SELECT 4
Time: 0.002s
```

Here is how my `pip freeze` looks. See if anything is different with yours:

```
(pgcli36b2) --- src/pgcli ‹master*?› » pip freeze
click==6.6
configobj==5.0.6
humanize==0.5.1
-e git+git@github.com:dbcli/pgcli.git@10a4f9c46ab9dc07dac5b2d58b9465f88625f095#egg=pgcli
pgspecial==1.6.0
prompt-toolkit==1.0.8
psycopg2==2.6.2
Pygments==2.1.3
setproctitle==1.1.10
six==1.10.0
sqlparse==0.2.2
wcwidth==0.1.7
```

**Comment by lelit** [comment]
Very strange indeed! I just created a new virtualenv using exactly your pinnings, and a space after `\i` immediately triggers the problem.
I will try to investigate further on my part and let you know.
BTW, do you know if there is some trick to debug the tool with `pdb`? When entering an interactive debug session, there seems to be a fight between the `pgcli` input routines and `pdb` ones, contending the same stdin stream...

**Comment by lelit** [comment]
Inserting a `print(matches)` when the exception happens in pgcompleter.py, I see the following:
```
[Match(completion=Completion(text='bin', start_position=0, display='bin/'), priority=None), Match(completion=Completion(text='include', start_position=0, display='include/'), priority=None), Match]
```
so effectively the `priority` is `None`, and that's what triggers the problem in the expression `sorted(matches, key=operator.attrgetter('priority'), reverse=True)`

**Comment by j-bennet** [comment]
@lelit That's something! For debugging, I usually do it on a checkout of pgcli code. Then inserting `import ipdb; ipdb.set_trace()` works just fine. If you feel like exploring it by yourself, PRs are always welcome! To me, this looks like a culprit:

https://github.com/dbcli/pgcli/blob/master/pgcli/pgcompleter.py#L640

it actually yields matches with `None`. I have no idea how this does not break things for me.

**Comment by j-bennet** [comment]
For the record, here is the difference between python 2.7 and 3.6 when sorting `None`.

2.7:

```
Python 2.7.12 (default, Oct 20 2016, 10:17:35)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.38)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> sorted([None, None, None])
[None, None, None]
```

3.6:

```
Python 3.6.0b2 (default, Oct 30 2016, 15:05:01)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> sorted([None, None, None])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'NoneType' and 'NoneType'
```

**Comment by lelit** [comment]
Yes, that seems the culprit. Replacing that line with `yield Match(completion=c, priority=0)` fixes the issue.
The [previous version](https://github.com/dbcli/pgcli/commit/f83044f2f0f30e872471c63eca4844511ca7ae93#diff-36fcb873925169e72e8c5ca4e8a8a581L385) of that code used the underlying `find_matches()` to actually build the result, and that always (not checked all the code paths, but seems so) inserts a _not_ `None` priority.

**Comment by lelit** [comment]
Also [this line](https://github.com/dbcli/pgcli/blob/master/pgcli/pgcompleter.py#L412) is suspicious: in general, under Python 3 you cannot compare different types, so even that case may generate the error
```
Python 3.6.0b2+ (default, Oct 31 2016, 08:51:43) 
[GCC 6.2.0 20161027] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> sorted([1, (1,1,1)])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'tuple' and 'int'
```

**Comment by j-bennet** [comment]
Yes, I think we should always make priority a tuple. In this case, `(0,)`

**Comment by lelit** [comment]
Ok, working on a PR right now.

**Comment by j-bennet** [comment]
🍫

**Comment by j-bennet** [comment]
https://github.com/dbcli/pgcli/issues/603 was merged.

**Comment by tcaruso42** [comment]
Hi,

I'm getting this same error using pgcli 1.3.1 on python 3.4.3. immediately after typing `\i ` I get:

```
\i Exception in thread Thread-23:
Traceback (most recent call last):                                                                       
  File "/usr/local/lib/python3.4/threading.py", line 920, in _bootstrap_inner                            
    self.run()                                                                                           
  File "/usr/local/lib/python3.4/threading.py", line 868, in run                                         
    self._target(*self._args, **self._kwargs)                                                            
  File "/usr/local/lib/python3.4/site-packages/prompt_toolkit/interface.py", line 840, in run            
    completions = list(buffer.completer.get_completions(document, complete_event))                       
  File "/usr/local/lib/python3.4/site-packages/pgcli/pgcompleter.py", line 371, in get_completions       
    reverse=True)                                                                                        
TypeError: unorderable types: NoneType() < NoneType()
```

Was this fix officially released, or would I need to install from the github master?

Thanks,

Tom

**Comment by j-bennet** [comment]
There was no release after that merge. I'm planning a new release after https://github.com/dbcli/pgcli/pull/624 is merged, but for now, install from master.

**Comment by amjith** [comment]
You can install from master using this command: 

```
pip install -U https://github.com/dbcli/pgcli/archive/master.zip
```

**Comment by j-bennet** [comment]
Fixed in 1.4.0.
